### ISSUE TITLE
Inheritance of optional URL parameters not working

### ISSUE THREAD

**Issue by lermontex** [vuejs/router#1682]
### Version
4.1.6

### Reproduction link
[codesandbox.io](https://codesandbox.io/p/sandbox/blissful-kowalevski-kotuh6)

### Steps to reproduce
1. Go to /en/:

```
https://kotuh6-5173.preview.csb.app/en/
```

2. Explore Navigation Links:

```
https://kotuh6-5173.preview.csb.app/en/
https://kotuh6-5173.preview.csb.app/en/about
```

3. Select German language
4. Explore Navigation Links:

```
https://kotuh6-5173.preview.csb.app/de/
https://kotuh6-5173.preview.csb.app/de/about
```

5. Make the `locale` parameter optional:

```
"/:locale(de|en)" ---> "/:locale(de|en)?" (vite.config.js)
```

6. Go to /en/:

```
https://kotuh6-5173.preview.csb.app/en/
```

7. Explore Navigation Links:

```
https://kotuh6-5173.preview.csb.app/
https://kotuh6-5173.preview.csb.app/about
```

8. Select German language
9. Explore Navigation Links:

```
https://kotuh6-5173.preview.csb.app/
https://kotuh6-5173.preview.csb.app/about
```

### What is expected?
I expect the `locale` parameter to be populated if passed in the URL

### What is actually happening?
If the `locale` parameter is not required, it is not filled in even if passed in the URL

---
Perhaps I am doing something wrong? How can I fill in the `locale` parameter, automatically?

I understand that I can pass it manually:

```
<router-link
    :to="{
        name: 'index', params: { locale: $route.params.locale }
    }"
>
Home
</router-link>
```

Or use `beforeEach` to fetch the parameter from the previous route (in this case the locale will not be visible in the url and it doesn't make sense if I use SSR)

But, isn't there another way?

**Comment by posva** [comment]
This is because the parameter is optional so not including it is the way to remove it. So yes, you have to manually pass it. Using a `beforeEach()` should also be a way to ensure the parameter is there:

```ts
router.beforeEach((to, from) => {
  if (!to.params.locale && from.params.locale) {
    return { ...to, params: { ...to.params, locale: from.params.locale }}
  }
})
```

Only required params are automatically preserved between navigations when the parameter matches but maybe this shouldn't be the default? I think the idea goes way beyond this though and has a big impact on the usage of vue router, so that should go in a proper RFC

**Comment by lermontex** [comment]
@posva,

> I think the idea goes way beyond this though and has a big impact on the usage of vue router, so that should go in a proper RFC

I opened a new Issue #1697 where I suggested a solution to this problem, but it was closed as a duplicate

What you suggest does not solve the SSR problem. The `beforeEach` hook will work after the user clicks on the link, and in the code there will be links like `example.com/page` instead of `example.com/en/page`

Once again, after rendering the HTML (renderToString), the page will have links regardless of the locale (if this is an optional parameter)

For example, the Google crawler will see that the links lead to `example.com/page` and not `example.com/en/page`

The `locale` example is just an example. I don't understand why you think this issue is outside the scope of Vue Router.
